### ISSUE TITLE
Ruby: Zactors can't be destroyed at process exit

### ISSUE THREAD

**Issue by paddor** [zeromq/czmq#1274]
I've been having this issue for quite a few days now and just can't figure out the reason why.

As far as I can see, I've implemented the creation of Zactors correctly, including the handshake. Automatic destroying of the Zactor does happen at process exit, as done by the boilerplate code from the generated Ruby binding.

Unfortunately, in almost all cases, it hangs when trying to destroy the leftover Zactors, until I send SIGTERM manually from another terminal. In that case, the process exits with a message like this:

```
E: 15-12-24 11:23:35 dangling 'PAIR' socket created at src/zsys.c:379
E: 15-12-24 11:23:35 dangling sockets: cannot terminate ZMQ safely
```

I stripped the code down to the minimum to show the issue. Most of it is boilerplate code copied right out of the generated Ruby binding. See [this gist](https://gist.github.com/paddor/4ad4e2452dd489791cf7).

Note that, when enabled, all the tests at the bottom pass before the issue occurs before process exit.

Requirements to be able to run the script:

* libzmq (tested with 4.1.3)
* libczmq (tested with 3.0.2)
* Ruby (tested with 2.2.4)
* FFI (tested with 1.9.10)
* RSpec (tested with 3.4.0)
* Minitest (tested with 5.8.3)

Thanks for any help. I'm about to go crazy.

**Comment by Asmod4n** [comment]
You have to disable the signal handling of czmq and trap SIGINT in ruby which sets zsys_interrupted to 1 i believe

**Comment by Asmod4n** [comment]
Take a look at https://github.com/Asmod4n/ruby-ffi-czmq/blob/master/examples/websocket_server.rb and https://github.com/Asmod4n/ruby-ffi-czmq/blob/master/lib/czmq/zsys.rb

**Comment by paddor** [comment]
Thanks so much! I'll give it a try...

**Comment by paddor** [comment]
Looks a lot better now using `zsys_handler_set(nil)`. Now it just hangs after cleaning up everything, even though the `at_exit` callback has been called to set `zsys_interrupted = true`. At least it exits on ^C.

By the way, [here](https://github.com/Asmod4n/ruby-ffi-czmq/blob/master/lib/czmq/zsys.rb#L39) you're just setting two local variables, not the attached ones. I tested it:

```ruby
module M
  @@foo = :foo

  def self.foo=(value)
    warn "setting @@foo to #{value.inspect}"
    @@foo = value
  end

  at_exit do
    self.foo = :bar1 #=> warning above is printed
    foo = :bar2      #=> nothing is printed
  end
end
```

**Comment by paddor** [comment]
Is there a way to send an interrupt to yourself? To me, it seems like setting the variable `zsys_interrupted` (or `zctx_interrupted`) would only have an effect if there was some kind of loop that would keep checking their value.

Now it just hangs before process exit, until I hit ^C, which does make sense.

In other words: Is there a way to interrupt all on-going calls to `zmsg_recv ()`?

**Comment by hintjens** [comment]
Terminating the context ends all blocking calls.

**Comment by paddor** [comment]
Thank you @hintjens, that helps. But it still results in this error before it exits:

```
E: 15-12-24 14:17:35 dangling 'PAIR' socket created at src/zsys.c:390
E: 15-12-24 14:17:35 dangling sockets: cannot terminate ZMQ safely
```

According to the debug output, the block that is called by the handler (FFI::Function) given to zactor_new() returns BEFORE zactor_destroy() sends "$TERM" to the actor and thus never gets an answer. zactor_destroy() gets called by the finalizer, set up by the generated Ruby binding. And that one only gets called when the generated Ruby binding CZMQ::FFI::Zactor object goes out of scope, which, of course, is after the higher-level CZTop::Actor object goes out of scope, where the code block was defined.

If in the block's `ensure` block (like Java's `finally`, I think) I send a signal to the pipe, I don't even need to `zsys_shutdown ()`, nor do I need to `zsys_handler_set (NULL)`, but the above error is still printed.

**Comment by paddor** [comment]
Oh nice. If I just destroy the pipe in the block's `ensure` block, the error is gone. In that case, zactor_destroy () still sees that there is a pipe set ([here](https://github.com/zeromq/czmq/blob/master/src/zactor.c#L168-L170)), but at least realizes that "$TERM" couldn't be sent => no error anymore

**Comment by paddor** [comment]
I was too fast with closing this. It's still a race condition. Even if I send a signal down the pipe and then destroy it, zactor_destroy () is late enough so sending "$TERM" down the pipe doesn't work (good case), or it is too early and sending "$TERM" still worked, but there will never be an answer signal because the Ruby block is gone.

Using `zsys_shutdown ()` works in all cases, but still causes the following to be pritned:

```
E: 15-12-24 15:04:59 dangling 'PAIR' socket created at src/zsys.c:390
E: 15-12-24 15:04:59 dangling 'PAIR' socket created at src/zsys.c:391
E: 15-12-24 15:04:59 dangling 'PAIR' socket created at src/zsys.c:391
E: 15-12-24 15:04:59 dangling 'PAIR' socket created at src/zsys.c:390
E: 15-12-24 15:04:59 dangling 'PAIR' socket created at src/zsys.c:391
E: 15-12-24 15:04:59 dangling 'PAIR' socket created at src/zsys.c:391
E: 15-12-24 15:04:59 dangling sockets: cannot terminate ZMQ safely
Assertion failed: (rc == 0 || zmq_errno () == ETERM), function zsock_set_sndtimeo, file ./src/zsock_option.inc, line 1335.
```

**Comment by Asmod4n** [comment]
Couldn't reliably solve this in my bindings either, would be cool if you find a way.
In mruby this doesn't work at all, since you cannot pass a block to another thread. Wrote https://github.com/Asmod4n/mruby-actor for that.

In CRuby it doesn't help with concurrency, since you can only run one ruby method at all times.

**Comment by paddor** [comment]
Well if you can't do it, how am I supposed to do it? :laughing: 

Can't there be something like a cleanup handshake? The initialization handshake works very reliable, after all.

Or maybe it just means that I have to pass the block to a method on CZMQ::FFI::Zactor, because that objet will outlive the associated CZTop::Actor object...

**Comment by hintjens** [comment]
Are you making a pure Ruby implementation of actors? Not a wrapping of CZMQ?

You might want to look at the zactor shim, which does the necessary
blocking on the pipe sockets:

    zsock_set_sndtimeo (shim->pipe, 0);
    zsock_signal (shim->pipe, 0);
    zsock_destroy (&shim->pipe);

Without this, you can't shutdown reliably. Note the zsock_signal sends a
signal back to the destructor in the API thread which is waiting for an
answer to $TERM.

-Pieter

**Comment by paddor** [comment]
Not pure Ruby, it's a wrapping for zactor.

Thanks, I was just reading those lines. The problem is that the handler doesn't return in that case. I just verified. It seems like the FFI::Function doesn't actually return, even though the block given to it is aborted by Ruby (probably when GC'ing the object). So the signal from the shim isn't executed.

Could this be a bug in Ruby-FFI? Or Ruby? :S

**Comment by hintjens** [comment]
So perhaps the actor destructor is being called in shutdown code that
doesn't wait? If for instance the zmq context is terminated, the actor will
not wait for the signal. Then you can see the socket leaks (as the actor
thread is still running when the process ends).

**Comment by jemc** [comment]
@paddor - I think you'll find that the order of garbage collection of interdependent objects in Ruby is rather undefined.  That is, if you have a group of objects that refer to each other in some graph, but you drop all "outside" references to these objects so that they are no longer reachable in the program, they will be garbage collected in some arbitrary order, no matter how they are connected by references to each other.

As you mentioned, the Ruby block can be garbage collected before the ZActor object (which runs the `zactor_destroy` finalizer).  This is not something I think Ruby with "fix" anytime soon, as the order is considered undefined behavior.  To make it worse, I think it's likely that even if you figure out how to work with the MRI Ruby behavior (which is not formally defined), it could change in a future version, and it won't work the same with Rubinius and JRuby.

In short, I think your only safe option here is to destroy explicitly, instead of relying on the GC order to be compatible with the needs of the finalizer.

**Comment by Asmod4n** [comment]
Afaik the only guarantee you get about "GC" order in ruby is how the ```at_exit``` blocks are called, they are called in the order they are created.

**Comment by paddor** [comment]
Thanks for the input. I just verified that the FFI callback function doesn't return in this case. See [this gist](https://gist.github.com/paddor/3cb70f285552ba65a620). It seems to behave the same on MRI, JRuby and Rubinius. 

Seems like anybody who uses zactors via CZTop really has to explicitly destroy them.

**Comment by paddor** [comment]
It works pretty well now. In the `ensure` block, I now start a thread whose sole purpose is to wait for the handler thread to die, because that's when the handler definitely has terminated, right? Then it'll inform the waiting `#terminate` method. It does this not with a Ruby ConditionVariable like I first tried, because the side thread might send the signal before the main thread starts to wait (which will result in a `fatal` exception in case there is no other thread, or a hang). Instead, it just uses a Ruby Queue and send `nil`. Kinda hacky, but reliable.

It also seems to work well in case one doesn't explicitly `#terminate` the actor, although I'm not sure why. But having thought about it again, this really is an edge case. I strongly assume one usually does not want to exit the process when actors are still running.
